src/controlplane/client/connector.go

```diff
@@ -0,0 +1,274 @@
+}
+
+type defaultConnector struct {
+	// TODO: Use a non-sdk logger
+	logger          log.Logger
+	kc              kubernetes.Interface
+	inClusterConfig *rest.Config
+	endpoints       []config.Endpoint
+}
+
+// DefaultConnector returns a defaultConnector that probes all endpoints in the list and return the first responding status OK.
+func DefaultConnector(endpoints []config.Endpoint, kc kubernetes.Interface, inClusterConfig *rest.Config, logger log.Logger) (Connector, error) {
+	if err := validateEndpointConfig(endpoints); err != nil {
+		return nil, fmt.Errorf("validating endpoints config: %w", err)
+	}
+
+	return &defaultConnector{
+		logger:          logger,
+		inClusterConfig: inClusterConfig,
+		kc:              kc,
+		endpoints:       endpoints,
+	}, nil
+}
+
+// Connect iterates over the endpoints list and returns the connection parameters of the
+// first endpoint that respond Status OK.
+func (dp *defaultConnector) Connect() (*connParams, error) {
+	for _, e := range dp.endpoints {
+
+		u, err := url.Parse(e.URL)
+		if err != nil {
+			return nil, fmt.Errorf("parsing endpoint url %q: %w", e.URL, err)
+		}
+
+		// If no path is defined on the config, default is set.
+		if u.Path == "" || u.Path == "/" {
+			u.Path = defaultMetricsPath
+		}
+
+		httpClient, err := dp.newHTTPClient(e)
+		if err != nil {
+			return nil, fmt.Errorf("creating HTTP client for endpoint %q: %w", e.URL, err)
+		}
+
+		if err := dp.probeEndpoint(u.String(), httpClient); err != nil {
+			dp.logger.Debugf("Autodiscover endpoint %q probe failed: %v", e.URL, err)
+			continue
+		}
+
+		// First endpoint with OK probe results returns.
+		return &connParams{url: *u, client: httpClient}, nil
+	}
+
+	return nil, fmt.Errorf("all autodiscover endpoints probe failed to response")
+}
+
+func (dp *defaultConnector) probeEndpoint(url string, client *http.Client) error {
+	resp, err := client.Head(url)
+	if err != nil {
+		return fmt.Errorf("http request failed with error: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http request failed with status: %v", resp.Status)
+	}
+
+	return nil
+}
+
+func validateEndpointConfig(endpoints []config.Endpoint) error {
+	for _, e := range endpoints {
+
+		if _, err := url.Parse(e.URL); err != nil {
+			return fmt.Errorf("parsing endpoint url %q: %w", e.URL, err)
+		}
+
+		if err := validateAuth(e.Auth); err != nil {
+			return fmt.Errorf("validating auth for endpoint url %q: %w", e.URL, err)
+		}
+	}
+
+	return nil
+}
+
+func validateAuth(auth *config.Auth) error {
+	if auth == nil {
+		return nil
+	}
+
+	switch {
+	case auth.Type == plainAuth:
+		break
+	case strings.EqualFold(auth.Type, bearerAuth):
+		break
+	case strings.EqualFold(auth.Type, mTLSAuth):
+		return validateMTLS(auth.MTLS)
+	default:
+		return fmt.Errorf("authorization type not supported: %q", auth.Type)
+	}
+
+	return nil
+}
+
+func validateMTLS(mTLS *config.MTLS) error {
+	if mTLS == nil {
+		return fmt.Errorf("mTLS config must exist")
```

Comment:
In the cases where a new error is created, I prefer to create a custom Error at the top of the file like errConfigMustExist = errors.new("mTLS config must exist") and return it here. Then, it can be checked afterwards with errors.is(errConfigMustExist), for example on a unitTest.

Reply:
oks, yes, for sure make it the same approach everywhere is the best.

Reply:
Maybe can be discussed for a final refactor.